source/particle/property_pool.cc

```diff
@@ -0,0 +1,100 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/particle/property_pool.h>
+#include <aspect/particle/particle.h>
+
+#include <aspect/global.h>
+#include <aspect/utilities.h>
+#include <aspect/compat.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    PropertyPool::PropertyPool (const unsigned int n_properties_per_slot)
+    :
+          n_properties (n_properties_per_slot),
+          memory_pool(),
+          translation_table(),
+          free_slots()
+    {}
+
+    PropertyPool::Handle
+    PropertyPool::allocate_properties_array ()
+    {
+      if (n_properties == 0)
+        return numbers::invalid_unsigned_int;
+
+      if (free_slots.is_empty())
+        reserve(free_slots.size() * 2 + 1);
+
+      const Handle handle = free_slots.pop_back();
+      translation_table[handle] = handle * n_properties;
+
+      return handle;
+    }
+
+    void
+    PropertyPool::deallocate_properties_array (Handle handle)
+    {
+      free_slots.add_index(handle);
```

Comment:
Ah, I think I know what you are thinking of (or not, but something else that might help). Since we organized the properties in a way that we can move them around, we can swap the properties of the particle that is going to be deleted with the ones from the particle next to a free range (e.g. the large free range at the end of the IndexSet) and then release the properties without creating a new range in the IndexSet. Is this approximately what you were thinking of?